Add initial CircleCI support

stellarspot opened the pull request:
Add CircleCI support to build and test platform-contracts project.

vsbogd commented on .circleci/config.yml:
```diff
@@ -0,0 +1,19 @@
+    - run:
+        name: Build platform-contracts
+        command: |
+          npm install ganache-cli
```
ganache-cli should not be necessary to build and test platform-contracts solely

stellarspot replied:
removing ganache-cli installation leads to the error: Cannot find module 'ethereumjs-abi'

vsbogd replied:
It should be installed by ```npm install``` as transitive dependency:
```
singularitynet-platform-contracts@0.2.1 /home/vital/projects/singnet/src/github.com/singnet/platform-contracts
│─┬ truffle-hdwallet-provider@0.0.5 
│ └─┬ web3-provider-engine@14.0.6
│   ├─┬ eth-sig-util@1.4.2 
│   │ └── ethereumjs-abi@0.6.5  (git+https://github.com/ethereumjs/ethereumjs
```

stellarspot replied:
I had: UNMET DEPENDENCY ethereumjs-abi

├─┬ truffle-hdwallet-provider@0.0.5
│ └─┬ web3-provider-engine@14.0.6
│   ├─┬ eth-sig-util@1.4.2
│   │ ├── UNMET DEPENDENCY ethereumjs-abi@git+https://github.com/ethereumjs/ethereumjs-abi.git
│   │ └─┬ ethereumjs-util@5.2.0

tiero replied:
Can you try install it as devDependency? (ganache-cli I mean)

vsbogd replied:
Adding ganache-cli as dev dependency may be useful to not install it each time. 
But the question is why one npm instance resolves dependencies correctly and another doesn't.

stellarspot replied:
The unmet dependencies with ethereumjs-abi are reproduced with npm 5.6.0 which is shipped with the docker image circleci/node:8.10.
I tried the latest npm 6.4.1 on the same environment and ethereumjs-abi is properly installed.

vsbogd commented on .circleci/config.yml:
```diff
@@ -0,0 +1,19 @@
+        command: |
+          npm install ganache-cli
+          npm install
+          npm run-script compile
```
1. ```npm run test``` actually does ```rm -rf build``` before running ```truffle test```: https://github.com/singnet/platform-contracts/blob/4ef66e32020af51133ca80d79d39b1e025328898/package.json#L13
2. ```truffle test``` always run compilation before testing

Taking this into account I would propose using ```npm run-script test``` instead of ```npm run-script compile``` to no build contracts twice.

stellarspot replied:
I would prefer to have unified way of compiling and testing projects.
In this way it is better to have separate compile and test steps.
I can update the test script to rid of build folder removing.

stellarspot replied:
The truffle itself compiles tests and contracts. I will remove the compilation and package steps from the CircleCI script.

tiero commented:
ACK

stellarspot commented:
I updated the fix to use circleci/node:10.11 docker image to use npm 6.4.1 which properly install ethereumjs-abi from dependencies.
The docker image also contains node v10.11.0 which now fails to compile test/util/CollectionExtensions.js:
```
/home/circleci/singnet/platform-contracts/test/util/CollectionExtensions.js:52
    return this.map((e,i) => [this[i], i < otherArray.length ? otherArray[i] : null]);
                ^
TypeError: this.map is not a function```

I updated  zip and difference methods defenition to use Object.defineProperty(...).

tiero commented:
Just from my heuristic observation in the past, Linux and node/npm do not work well every time and this is one of the reasons Facebook did `yarn`, which solves some glitches on linux machines. I have no facts/issues to demonstrate it. Give it a try if you can't find the solution in other ways
